Program.cs

```diff
@@ -13,23 +15,141 @@ private static bool isDivisibleBy(int number, int divisior)
         return number % divisior == 0;
     }
 
-    private static void responseToNumber(int number)
+    // order of the action matters, they must be applied in this order
+    private enum CompositionActions
     {
-        if (isDivisibleBy(number, 3) && isDivisibleBy(number, 5))
-        {
-            Console.WriteLine("FizzBuzz");
-        }
-        else if (isDivisibleBy(number, 5)) 
+        AppendFizz,
+        AppendBuzz,
+        AppendBang,
+        ReplaceEverythingWithBong,
+        SuffixWithFezz,
+        ReverseAll
+    }
+
+    private static string responseToNumber(int number)
+    {
+        var divisorsAndNames = new Dictionary<int, CompositionActions>
         {
-            Console.WriteLine("Buzz");
-        }
-        else if (isDivisibleBy(number, 3))
+            { 3, CompositionActions.AppendFizz },
+            { 5, CompositionActions.AppendBuzz },
+            { 7, CompositionActions.AppendBang },
+            { 11, CompositionActions.ReplaceEverythingWithBong },
+            { 13, CompositionActions.SuffixWithFezz },
+            { 17, CompositionActions.ReverseAll }
+        };
+
+        var answerComponents = new List<CompositionActions> { };
+        bool showNumber = true;
+
+        foreach (var divisor in divisorsAndNames.Keys)
+            if (isDivisibleBy(number, divisor))
+            {
+                var action = divisorsAndNames[divisor];
+                answerComponents.Add(divisorsAndNames[divisor]);
+                if (action != CompositionActions.ReverseAll)
+                    showNumber = false;
+            }
+
+        if (showNumber)
+            return number.ToString();
+        else
+            return composeString(answerComponents.ToArray());
+    }
+
+    private static string composeString(CompositionActions[] list)
+    {
+        var output = new List<string>();
+        int indexOfFirstB = -1;
+
+        foreach (var action in list)
         {
-            Console.WriteLine("Fizz");
+            switch (action)
+            {
+                case CompositionActions.AppendFizz:
+                    {
+                        output.Add("Fizz");
+                        break;
+                    }
+                case CompositionActions.AppendBuzz:
+                case CompositionActions.AppendBang:
+                    {
+                        string name = action.ToString().Replace("Append", "");

```

Comment:
This feels a little bit hacky, but I'm not really sure a better way to do it. I think this general approach of using "composition actions" feels like it could be simplified a bit by removing one layer of abstraction. Instead of having a switch statement you could have a series of if statements resolving in order?

There's also another interesting way you could use which involves making `CompositionActions` a class which can take a divisor and return what word should happen there? In this Java "model answer" you can see some cool different approaches:  https://github.com/CorndelWithSoftwire/fizzbuzz-java/blob/master/src/training/fizzbuzz/Part2.java

Reply:
I really like the idea of the `CompositionActions` as a class definitely would make the code more readable and allow me to read about OOP in C#, so I will look into doing that if I have some time today

Reply:
Decided to experiment with C# inheirtance and see how much I can learn on the way, think the current solution is much nicer to read, but still not perfect!